sdk/pubkey/src/lib.rs

```diff
@@ -73,39 +144,71 @@ impl From<u64> for PubkeyError {
 /// [`Keypair`]: https://docs.rs/solana-sdk/latest/solana_sdk/signer/keypair/struct.Keypair.html
 #[cfg_attr(target_arch = "wasm32", wasm_bindgen)]
 #[repr(transparent)]
-#[cfg_attr(feature = "frozen-abi", derive(AbiExample))]
+#[cfg_attr(feature = "frozen-abi", derive(solana_frozen_abi_macro::AbiExample))]
 #[cfg_attr(
     feature = "borsh",
-    derive(BorshSerialize, BorshDeserialize, BorshSchema),
+    derive(BorshSerialize, BorshDeserialize),
     borsh(crate = "borsh")
 )]
-#[derive(
-    Clone,
-    Copy,
-    Default,
-    Deserialize,
-    Eq,
-    Hash,
-    Ord,
-    PartialEq,
-    PartialOrd,
-    Pod,
-    Serialize,
-    Zeroable,
-)]
-#[cfg_attr(test, derive(Arbitrary))]
+#[cfg_attr(all(feature = "borsh", feature = "std"), derive(BorshSchema))]
+#[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
+#[cfg_attr(feature = "bytemuck", derive(Pod, Zeroable))]
+#[derive(Clone, Copy, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+#[cfg_attr(any(test, feature = "dev-context-only-utils"), derive(Arbitrary))]
 pub struct Pubkey(pub(crate) [u8; 32]);
 
 impl solana_sanitize::Sanitize for Pubkey {}
 
-#[derive(Error, Debug, Serialize, Clone, PartialEq, Eq, FromPrimitive, ToPrimitive)]
+#[cfg_attr(feature = "serde", derive(Serialize))]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub enum ParsePubkeyError {
-    #[error("String is the wrong size")]
     WrongSize,
-    #[error("Invalid Base58 string")]
     Invalid,
 }
 
+impl ToPrimitive for ParsePubkeyError {
+    #[inline]
+    fn to_i64(&self) -> Option<i64> {
+        Some(match *self {
+            ParsePubkeyError::WrongSize => ParsePubkeyError::WrongSize as i64,
+            ParsePubkeyError::Invalid => ParsePubkeyError::Invalid as i64,
+        })
+    }
+    #[inline]
+    fn to_u64(&self) -> Option<u64> {
+        self.to_i64().map(|x| x as u64)
+    }
+}
+
+impl FromPrimitive for ParsePubkeyError {
+    #[inline]
+    fn from_i64(n: i64) -> Option<Self> {
+        if n == ParsePubkeyError::WrongSize as i64 {
+            Some(ParsePubkeyError::WrongSize)
+        } else if n == ParsePubkeyError::Invalid as i64 {
+            Some(ParsePubkeyError::Invalid)
+        } else {
+            None
+        }
+    }
+    #[inline]
+    fn from_u64(n: u64) -> Option<Self> {
+        Self::from_i64(n as i64)
+    }
+}
```

Comment:
Same as with the other `FromPrimitive` implementation, is there a way to add an exhaustiveness check? If not, would you mind adding a simple test to make sure that the conversion works? ie

```
        macro_rules! ensure_mapping {
            ($ty: path, $($num: literal => $variant: path),+ $(,)?) => {
                // assert that the given values produce the expected variants
                $(assert_eq!(<$ty>::from($num), $variant);)+
                $(assert_eq!(<$ty>::from_i64($num), Some($variant));)+
                $(assert_eq!(<$ty>::from_u64($num), Some($variant));)+

                // this generated fn will never be called but will produce a
                // non-exhaustive pattern error if you've missed a variant
                #[allow(dead_code)]
                fn check_all_covered(err: $ty) {
                    match err {
                        $($variant => {})+
                    };
                }
            }
        }
    #[test]
    fn test_exhaustive_conversion() {
        ensure_mapping! {
            PubkeyError,
            0 => PubkeyError::MaxSeedLengthExceeded,
            1 => PubkeyError::InvalidSeeds,
            2 => PubkeyError::IllegalOwner,
        }
        ensure_mapping! {
            ParsePubkeyError,
            0 => ParsePubkeyError::WrongSize,
            1 => ParsePubkeyError::Invalid,
        }
    }
```

You'll need to add a `From<u64>` implementation for `ParsePubkeyError` to get this to work, unless you have a more clever way to only do that check for the first enum

Reply:
I used `strum` to iterate through variants and check that strum's `from_repr` matches our `from_i64`:
```rust
for variant in PubkeyError::iter() {
    let variant_i64 = variant as i64;
    assert_eq!(
        PubkeyError::from_repr(variant_i64 as usize),
        PubkeyError::from_i64(variant_i64)
    );
}
```